omod/src/main/java/org/openmrs/module/reportingrest/web/resource/ReportRequestResource.java

```diff
@@ -65,27 +70,103 @@ public ReportRequest newDelegate() {
+		}
+
+		for (ReportRequest reportRequest : reportRequests) {
+			for (RenderingMode mode :
+					reportService.getRenderingModes(
+							reportRequest.getReportDefinition().getParameterizable())) {
+				if (OpenmrsUtil.nullSafeEquals(mode, reportRequest.getRenderingMode())) {
+					reportRequest.setRenderingMode(mode);
+				}
+			}
+		}
+
+		return new AlreadyPaged<ReportRequest>(context, reportRequests, reportRequestsTotalCount > (long) pageNumber * pageSize, reportRequestsTotalCount);
+	}
+
+	private List<ReportRequest.Status> findAppropriateStatuses(RequestContext context) {
+		List<ReportRequest.Status> reportRequestStatusList = new ArrayList<ReportRequest.Status>();
+		String commaSeparatedStatusesParam = context.getParameter("status");
+		if (StringUtils.isBlank(commaSeparatedStatusesParam)) {
+			return reportRequestStatusList;
+		}
+
+		String[] splitStatuses = commaSeparatedStatusesParam.split(",");
+		for (String status : splitStatuses) {
+			if ("ran".equalsIgnoreCase(status)) {
+				Collections.addAll(reportRequestStatusList,
+						ReportRequest.Status.REQUESTED,
+						ReportRequest.Status.COMPLETED,
+						ReportRequest.Status.SAVED,
+						ReportRequest.Status.FAILED,
+						ReportRequest.Status.PROCESSING);
+			} else if ("processing".equalsIgnoreCase(status)) {
+				Collections.addAll(reportRequestStatusList,
+						ReportRequest.Status.REQUESTED,
+						ReportRequest.Status.PROCESSING);
+			} else if ("scheduled".equalsIgnoreCase(status)) {
+				Collections.addAll(reportRequestStatusList,
+						ReportRequest.Status.SCHEDULED,
+						ReportRequest.Status.SCHEDULE_COMPLETED);
+			}
 		}
-		ReportDefinition reportDefinition = DefinitionContext.getDefinitionByUuid(ReportDefinition.class, reportDefinitionParam);
-		if (reportDefinition == null) {
-			throw new NullPointerException("Cannot find reportDefinition=" + reportDefinitionParam);
+
+		return reportRequestStatusList;
+	}
+
+	private ReportDefinition findReportDefinition(RequestContext context) {
+		ReportDefinition reportDefinition = null;
+		String reportDefinitionParam = context.getParameter("reportDefinition");
+		if (reportDefinitionParam != null) {
+			reportDefinition = DefinitionContext.getDefinitionByUuid(ReportDefinition.class, reportDefinitionParam);
 		}
-		List<ReportRequest> reportRequests = getService().getReportRequests(reportDefinition, null, null);
-		return new NeedsPaging<ReportRequest>(reportRequests, context);
+
+		return reportDefinition;
 	}
-	
+
 	/**
 	 * @see org.openmrs.module.webservices.rest.web.resource.impl.DelegatingResourceHandler#save(java.lang.Object)
 	 */
 	@Override
     public ReportRequest save(ReportRequest reportRequest) {
-		return getService().saveReportRequest(reportRequest);
+		ReportService reportService = getService();
+		if (reportService.getReportRequestByUuid(reportRequest.getUuid()) == null) {
+			return reportService.queueReport(reportRequest);
+		} else {
+			return reportService.saveReportRequest(reportRequest);
+		}

```

Comment:
Based on how `reportService.queueReport(ReportRequest)` is implemented, I think we should just be able to always call `queueReport` here, as all the `queueReport` method does is populate the status if it is null and then save it.

Per the comment earlier up in the PR, we should likely change this method to just theselines:
```
reportRequest = reportService.queueReport(reportRequest);
reportService.processNextQueuedReports();
return reportRequest;
```

Reply:
corrected